File: 10-extended-bst/code.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Struktur Node untuk BST
typedef struct BSTNode {
    int data;
    struct BSTNode* left;
    struct BSTNode* right;
} BSTNode;

// Fungsi untuk membuat node baru
BSTNode* createNode(int data) {
    BSTNode* newNode = (BSTNode*)malloc(sizeof(BSTNode));
    if (newNode == NULL) {
        printf("Memory allocation failed\n");
        exit(1);
    }
    newNode->data = data;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

// Fungsi untuk menyisipkan node ke dalam BST
BSTNode* insert(BSTNode* root, int data) {
    if (root == NULL) {
        return createNode(data);
    }
    
    if (data < root->data) {
        root->left = insert(root->left, data);
    } else if (data > root->data) {
        root->right = insert(root->right, data);
    }
    
    return root;
}

// Fungsi untuk mencari nilai minimum dalam BST
BSTNode* findMin(BSTNode* root) {
    if (root == NULL) {
        return NULL;
    } else if (root->left == NULL) {
        return root;
    } else {
        return findMin(root->left);
    }
}

// Fungsi untuk menghapus node dari BST
BSTNode* deleteNode(BSTNode* root, int key) {
    if (root == NULL) {
        return root;
    }
    
    if (key < root->data) {
        root->left = deleteNode(root->left, key);
    } else if (key > root->data) {
        root->right = deleteNode(root->right, key);
    } else {
        // Node dengan satu anak atau tanpa anak
        if (root->left == NULL) {
            BSTNode* temp = root->right;
            free(root);
            return temp;
        } else if (root->right == NULL) {
            BSTNode* temp = root->left;
            free(root);
            return temp;
        }
        
        // Node dengan dua anak
        BSTNode* temp = findMin(root->right);
        root->data = temp->data;
        root->right = deleteNode(root->right, temp->data);
    }
    
    return root;
}

// Fungsi untuk traversal inorder
void inorder(BSTNode* root) {
    if (root != NULL) {
        inorder(root->left);
        // printf("%d ", root->data);
        inorder(root->right);
    }
}

// Fungsi untuk traversal preorder
void preorder(BSTNode* root) {
    if (root != NULL) {
        // printf("%d ", root->data);
        preorder(root->left);
        preorder(root->right);
    }
}

// Fungsi untuk traversal postorder
void postorder(BSTNode* root) {
    if (root != NULL) {
        postorder(root->left);
        postorder(root->right);
        // printf("%d ", root->data);
    }
}

// Fungsi untuk mencari nilai dalam BST
BSTNode* search(BSTNode* root, int key) {
    if (root == NULL || root->data == key) {
        return root;
    }
    
    if (key < root->data) {
        return search(root->left, key);
    }
    
    return search(root->right, key);
}

// Fungsi untuk membersihkan BST
void freeBST(BSTNode* root) {
    if (root != NULL) {
        freeBST(root->left);
        freeBST(root->right);
        free(root);
    }
}

int main() {
    clock_t start, end;
    double cpu_time_used;
    
    // Inisialisasi BST
    BSTNode* root = NULL;
    
    start = clock();
    
    // Operasi pada BST
    // Menambahkan node (misal 1000 node)
    for (int i = 0; i < 1000; i++) {
        int value = rand() % 10000;
        root = insert(root, value);
    }
    
    // Traversal
    inorder(root);
    preorder(root);
    postorder(root);
    
    // Menghapus beberapa node
    for (int i = 0; i < 100; i++) {
        int key = rand() % 10000;
        root = deleteNode(root, key);
    }
    
    // Mencari beberapa nilai
    for (int i = 0; i < 100; i++) {
        int key = rand() % 10000;
        BSTNode* result = search(root, key);
        if (result != NULL) {
            printf("Found %d\n", key);
        } else {
            printf("%d not found\n", key);
        }
    }
    
    freeBST(root);
    
    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    printf("Waktu eksekusi: %f detik\n", cpu_time_used);
    
    return 0;
}

File: 10-extended-bst/code.rs
use std::time::Instant;

// Struktur Node untuk BST
struct Node {
    data: i32,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

// Implementasi untuk Node
impl Node {
    // Membuat node baru
    fn new(data: i32) -> Self {
        Node {
            data,
            left: None,
            right: None,
        }
    }
}

// Struktur untuk mengelola BST
struct BST {
    root: Option<Box<Node>>,
}

// Implementasi untuk BST
impl BST {
    // Membuat BST baru
    fn new() -> Self {
        BST {
            root: None,
        }
    }
    
    // Menyisipkan node ke dalam BST
    fn insert(&mut self, data: i32) {
        self.root = Self::insert_recursive(self.root.take(), data);
    }
    
    // Helper function untuk insert
    fn insert_recursive(node: Option<Box<Node>>, data: i32) -> Option<Box<Node>> {
        match node {
            None => Some(Box::new(Node::new(data))),
            Some(mut node) => {
                if data < node.data {
                    node.left = Self::insert_recursive(node.left.take(), data);
                } else if data > node.data {
                    node.right = Self::insert_recursive(node.right.take(), data);
                }
                Some(node)
            }
        }
    }
    
    // Menghapus node dari BST
    fn delete(&mut self, key: i32) {
        self.root = Self::delete_recursive(self.root.take(), key);
    }
    
    // Helper function untuk delete
    fn delete_recursive(node: Option<Box<Node>>, key: i32) -> Option<Box<Node>> {
        match node {
            None => None,
            Some(mut node) => {
                if key < node.data {
                    node.left = Self::delete_recursive(node.left.take(), key);
                    Some(node)
                } else if key > node.data {
                    node.right = Self::delete_recursive(node.right.take(), key);
                    Some(node)
                } else {
                    // Node tanpa anak
                    if node.left.is_none() && node.right.is_none() {
                        None
                    }
                    // Node dengan satu anak (kanan)
                    else if node.left.is_none() {
                        node.right
                    }
                    // Node dengan satu anak (kiri)
                    else if node.right.is_none() {
                        node.left
                    }
                    // Node dengan dua anak
                    else {
                        let min_value = Self::find_min(&node.right);
                        node.data = min_value;
                        node.right = Self::delete_recursive(node.right.take(), min_value);
                        Some(node)
                    }
                }
            }
        }
    }
    
    // Mencari nilai minimum dalam BST
    fn find_min(node: &Option<Box<Node>>) -> i32 {
        match node {
            None => panic!("Empty tree"),
            Some(node) => {
                match &node.left {
                    None => node.data,
                    Some(_) => Self::find_min(&node.left),
                }
            }
        }
    }
    
    // Traversal inorder
    fn inorder(&self) {
        Self::inorder_recursive(&self.root);
    }
    
    // Helper function untuk inorder
    fn inorder_recursive(node: &Option<Box<Node>>) {
        if let Some(node) = node {
            Self::inorder_recursive(&node.left);
            // print!("{} ", node.data);
            Self::inorder_recursive(&node.right);
        }
    }
    
    // Traversal preorder
    fn preorder(&self) {
        Self::preorder_recursive(&self.root);
    }
    
    // Helper function untuk preorder
    fn preorder_recursive(node: &Option<Box<Node>>) {
        if let Some(node) = node {
            // print!("{} ", node.data);
            Self::preorder_recursive(&node.left);
            Self::preorder_recursive(&node.right);
        }
    }
    
    // Traversal postorder
    fn postorder(&self) {
        Self::postorder_recursive(&self.root);
    }
    
    // Helper function untuk postorder
    fn postorder_recursive(node: &Option<Box<Node>>) {
        if let Some(node) = node {
            Self::postorder_recursive(&node.left);
            Self::postorder_recursive(&node.right);
            // print!("{} ", node.data);
        }
    }
    
    // Mencari nilai dalam BST
    fn search(&self, key: i32) -> bool {
        Self::search_recursive(&self.root, key)
    }
    
    // Helper function untuk search
    fn search_recursive(node: &Option<Box<Node>>, key: i32) -> bool {
        match node {
            None => false,
            Some(node) => {
                if node.data == key {
                    true
                } else if key < node.data {
                    Self::search_recursive(&node.left, key)
                } else {
                    Self::search_recursive(&node.right, key)
                }
            }
        }
    }
}

fn main() {
    let start = Instant::now();
    
    // Inisialisasi BST
    let mut bst = BST::new();
    
    // Menambahkan node (misal 1000 node dengan nilai deterministik)
    for i in 0..1000 {
        let value = (i * 83) % 10000;
        bst.insert(value);
    }
    
    // Traversal
    bst.inorder();
    bst.preorder();
    bst.postorder();
    
    // Menghapus beberapa node
    for i in 0..100 {
        let key = (i * 97) % 10000;
        bst.delete(key);
    }
    
    // Mencari beberapa nilai
    for i in 0..100 {
        let key = (i * 101) % 10000;
        let found = bst.search(key);
        if found {
            // println!("Found {}", key);
        } else {
            // println!("{} not found", key);
        }
    }
    
    let duration = start.elapsed();
    println!("Waktu eksekusi: {:?}", duration);
}